Cascade layers post (#2044)
GoogleChrome/developer.chrome.com: site/en/blog/cascade-layers/index.md

    </figcaption>
</figure>

This example showcases the power of cascade layers, using `@layer`. There are several links shown: some without any additional class names applied, one with a `.link` class, and one with a `.pink` class. The CSS then adds  three layers: `base`, `typography`, and `utilities` as follows:

```css
@layer base {
  a {
    font-weight: 800;
    color: red; /* ignored */
  }

  .link {
    color: blue; /* ignored */
  }
}

@layer typography {
  a {
    color: green; /* styles *all* links */
  }
}

@layer utilities {
  .pink {
    color: hotpink;  /* styles *all* .pink's */
  }
}
```

Ultimately, all the links are either green or pink. This is because: while `.link` has a higher selector-level specificity than `a`, there is a color style on `a` in a higher-specificity `@layer`. `a { color: green }` overrides `.link { color: blue }` when the green rule is in a layer after the blue rule.

> The layer specificity beats the element specificity.

## Organizing layers

You can organize layers directly on the page, as shown above, or you can organize them at the top of a file. 

> > Layer order is established by the first time each layer name appears in your code.

Author: Yeah the idea was to increase visual hierarchy of these points. How would you recommend doing that?

       utilities;
```

Now, the order in which you `@import` your styles won’t matter to the layer order, since it’s already established at the first instance of the layer name. That's one less thing to worry about. You can still set imported files to specific layers, but the ordering is already established.

<figure>
    {% Img src="image/HodOHWjMnbNw56hvNASHWSgZyAf2/CSUb8jZLi4QJUdmTM109.png", alt="Screenshot from Codepen Project", width="800", height="556" %}
    <figcaption>Explore the <a href="https://codepen.io/web-dot-dev/project/editor/ZGQLkq">project on Codepen.</a>
    </figcaption>
</figure>

{% Aside %}
Importing styles in a CSS file like this chain as they load, so it's more performant to include everything in a `style` tag within your HTML. Stylesheets included via `link` are the most performant, but don't currently have the ability to load as layers. This is currently in discussion with the WHATWG.
{% endAside %}

## Layers and the cascade

Let’s take a step back and see where layers are used as it relates to the wider cascade:

{% Img src="image/HodOHWjMnbNw56hvNASHWSgZyAf2/jH8ydtcwmHg4rHTfqvRU.png", alt="Cascade Illustration", width="800", height="393" %}

> **UA normal < Local User @layer < Local User normal < Author @layers < Author normal < Author !important < Author @layer !important < Local User !important < UA !important**

Author: I want to put that in writing (i.e:

"UA normal has less precedence than Local User @layer which has less precedence than Local User normal which has less precedence than Author @layers which has less precedence than Author normal which has less precedence thanAuthor !important which has less precedence than Author @layer !important which has less precedence than Local User !important which has less precedence than UA !important"

> **UA normal < Local User @layer < Local User normal < Author @layers < Author normal < Author !important < Author @layer !important < Local User !important < UA !important**

Author: Maybe a list format is best here?

> You may notice here that `@layer !important` styles are inverted. Instead of being *less* specific than non-layered (normal) styles, they become *more* specific. This is because of how `!important` works in the cascade: it breaks the normal cascading in your stylesheets and reverses the normal specificity values.

Reviewer: ```suggestion
You may notice here that `@layer !important` styles are inverted. Instead of being less specific than non-layered (normal) styles, they become more specific. This is because of how `!important` works in the cascade: it breaks the normal cascading in your stylesheets and reverses the normal specificity values.
```

### Rule 1: Don’t use `@layer` for scoping

Cascade layers do not solve scoping. If you have  a CSS file with an @layer, say `card.css` and want to style all of the links within card, do *NOT* write styles like:

```css
a {
  …
}
```

This will lead to all of the `a` tags in your file getting this override. It’s still important to **scope** your styles properly:

```css
.card a {
  …
}
```

### Rule 2: cascade layers are ordered behind non-layered CSS

It’s important to note that a layered CSS file will *not* override non-layered CSS. This was an intentional decision to make it easier to introduce layers in a more sensible way to work with your existing codebase. Using a `reset.css` file, for example, is a good starting point and use case for cascade layers. 

### Rule 3: `!important` inverts cascade specificity

While layered styles are less specific than unlayered styles in general, using `!important` reverses this.

> Layers are less specific than unlayered styles *unless* they are `!important` styles within a layer.

> In that case, the `!important` styles invert their specificity. The diagram above shows this for reference: `Author @layers < Author normal < Author !important < Author @layer !important`.

Reviewer: can we write this one out too, rather than using `<`.

Author: this feels easier to read than `Author @layers have less precedence than Author normal which have less precedence than Author !important which have less precedence than Author @layer !important` but I'm happy to use either

@import '../styles/components/footer.css' layer(components); /* footer component */
```

The above code snippet has three layers: `base`,`layouts`, and `components`. The normalize, theme, and typography files in `base`, with a `post` file in `layouts`, and `cards` and `footer` both in `components`. On import of the file, the layers are instantiated using the layer function. An alternative approach would be to organize your layers at the top of the file, declaring them before any imports:

```css
@layer base,
       theme,
       layouts,
       components,
       utilities;
```

Now, the order in which you `@import` your styles won’t matter to the layer order, since it’s already established at the first instance of the layer name. That's one less thing to worry about. You can still set imported files to specific layers, but the ordering is already established.

<figure>
    {% Img src="image/HodOHWjMnbNw56hvNASHWSgZyAf2/CSUb8jZLi4QJUdmTM109.png", alt="Screenshot from Codepen Project", width="800", height="556" %}
    <figcaption>Explore the <a href="https://codepen.io/web-dot-dev/project/editor/ZGQLkq">project on Codepen.</a>
    </figcaption>
</figure>

{% Aside %}

> Importing styles in a CSS file like this chain as they load, so it's more performant to include everything in a `style` tag within your HTML. Stylesheets included via `link` are the most performant, but don't currently have the ability to load as layers. This is currently in discussion with the WHATWG.

Author: ```suggestion
Importing styles in a CSS file like this chain as they load, so it's more performant to include everything in a `style` tag within your HTML. Stylesheets included via `link` are the most performant, but don't currently have the ability to load as layers. This is currently [in discussion](https://github.com/whatwg/html/issues/7540) with the WHATWG.
```

> Importing styles in a CSS file like this chain as they load, so it's more performant to include everything in a `style` tag within your HTML. Stylesheets included via `link` are the most performant, but don't currently have the ability to load as layers. This is currently [in discussion](https://github.com/whatwg/html/issues/7540) with the WHATWG.

Author: ```suggestion
Importing styles in a CSS file like this chain as they load, so it's more performant to include everything in a `style` tag within your HTML. Stylesheets included via `link` are the most performant, but don't currently have the ability to load as layers. This is currently [an open issue](https://github.com/whatwg/html/issues/7540) with the WHATWG.
```
